src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,216 @@
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "PIXEL_SIZE" ), QObject::tr( "Pixel size" ), 10, QStringLiteral( "INPUT" ), false ) );
+
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Line density raster" ) ) );
+}
+
+QString QgsLineDensityAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "The line density interpolation algorithm calculates a density measure of linear features "
+                      "which is obtained in a circular neighborhood within each raster cell. "
+                      "First, the length of the segment of each line that is intersected by the circular neighborhood "
+                      "is multiplied with the lines weight factor. In a second step, all length values are summed and "
+                      "divided by the area of the circular neighborhood. This process is repeated for all raster cells."
+                    );
+}
+
+QgsLineDensityAlgorithm *QgsLineDensityAlgorithm::createInstance() const
+{
+  return new QgsLineDensityAlgorithm();
+}
+
+bool QgsLineDensityAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mWeightField = parameterAsString( parameters, QStringLiteral( "WEIGHT" ), context );
+
+  mPixelSize = parameterAsDouble( parameters, QStringLiteral( "PIXEL_SIZE" ), context );
+
+  mSearchRadius = parameterAsDouble( parameters, QStringLiteral( "RADIUS" ), context );
+  if ( mSearchRadius < std::sqrt( mPixelSize ) / 2 )
+    throw QgsProcessingException( QObject::tr( "Raster cells must be fully contained by the search circle. Therefore, "
+                                  "the search radius must not be smaller than half of the pixel diagonal (half of square root of pixel size)." ) );
+
+  mExtent = mSource->sourceExtent();
+  mCrs = mSource->sourceCrs();
+  mDa = QgsDistanceArea();
+
+  if ( context.project() )
+  {
+    mDa.setEllipsoid( context.project()->ellipsoid() );
+  }
+  mDa.setSourceCrs( mCrs, context.transformContext() );
+
+
+  //get cell midpoint from top left cell
+  QgsPoint firstCellMidpoint = QgsPoint( mExtent.xMinimum() + ( mPixelSize / 2 ), mExtent.yMaximum() - ( mPixelSize / 2 ) );
+  QgsCircle searchCircle = QgsCircle( firstCellMidpoint, mSearchRadius );
+  mSearchGeometry = QgsGeometry( searchCircle.toPolygon() );
+
+  return true;
+}
+
+QVariantMap QgsLineDensityAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mFeatureWeights = QHash<QgsFeatureId, double>();
```

Comment:
Not required

Reply:
I am a little confused here. In the previous review you advised me to 

> build up a hash of the weights vs feature id at the same time you populate the spatial index 

This sounds very reasonable to me because only the relevant QgsFeatureIds and respective weights stay in memory and no further calls to the provider are necessary in the loops below. I understand the inefficiency of inserting the default weight (=1) if no weight field is selected and the optimization via setSubsetOfAttributes(), but why shouldn't we use the QHash here? Maybe I am missing something ;)

Reply:
I just mean that particular line isn't required -- you don't need to manually initialise a qhash

Reply:
Thx! That was the thing I was missing. ;)